### PR TITLE
Use AddressSanitizer when available

### PR DESCRIPTION
**Pull request by Jules-Bertholet**
Compiles and runs the program and standard library with AddressSanitizer on targets that support it. This sanitizer ["is not expected to give false positive reports"](https://doc.rust-lang.org/beta/unstable-book/compiler-flags/sanitizer.html#addresssanitizer).

### CODE REVIEW

**Comment by RalfJung** [comment]
That sounds like a great idea, thanks!

Two things:

- I don't understand how sanitizers are used from rustc, or what half of this code does. Please add comments explaining that. Assume that the reader knows what a sanitizer is but never used one.
- What is the performance effect of this? Would it make sense to make this controllable via some user-exposed flag?

**Comment by Jules-Bertholet** [comment]
I've added some comments and a command-line flag to disable sanitizing. According to [clang docs](https://clang.llvm.org/docs/AddressSanitizer.html), slowdown is ~2x

File: src/main.rs [hunk 1]
```diff
@@ -233,14 +250,25 @@ fn build_sysroot(
...
 
     // Do the build.
-    eprint!("Preparing a careful sysroot (target: {target})... ");
+    eprint!(
+        "Preparing a careful sysroot (target: {target}{})... ",
+        if let Some(san) = sanitizer {
+            format!(", sanitizer: {san}")
+        } else {
+            String::new()
+        }
```
Comment (RalfJung) [inline src/main.rs:266]:
```suggestion
        if let Some(san) = sanitizer {
            format_args!(", sanitizer: {san}")
        } else {
            format_args!("")
        }
```
this should work just as well and avoids intermediate allocations

File: src/main.rs [hunk 2]
```diff
@@ -271,10 +299,19 @@ fn build_sysroot(
...
 
-    PathBuf::from(sysroot_dir)
+    sysroot_dir
 }
 
-fn cargo_careful(mut args: env::Args) {
+fn cargo_careful(args: env::Args) {
+    let mut args = args.peekable();
+
+    let use_sanitizer = if Some("--no-sanitizer") == args.peek().map(AsRef::as_ref) {
```
Comment (RalfJung) [inline src/main.rs:308]: I think we need to come up with some kind of concept for flags. I have an idea for that, but it needs a bit of design. I don't like mixing our flags with the cargo flag namespace and making the flag position-dependent is also rather ugly.

File: src/main.rs [hunk 3]
```diff
@@ -325,6 +374,12 @@ fn cargo_careful(mut args: env::Args) {
         "CARGO_ENCODED_RUSTDOCFLAGS",
         rustc_build_sysroot::encode_rustflags(&flags),
     );
+
+    // Leaks are not a memory safety issue, don't detect them by default
+    if sanitizer == Some("address") && !env::vars_os().any(|(key, _)| key == "ASAN_OPTIONS") {
```
Comment (RalfJung) [inline src/main.rs:379]: Why iterate over all env vars? `env::var_os("ASAN_OPTIONS).is_some()` makes a lot more sense, doesn't it?

File: src/main.rs (hunk 2 above)
Comment (RalfJung) [reply src/main.rs:303]: I plan to make a concrete suggestion tomorrow that I will push to your branch.

**Comment by thomcc** [comment]
I've used ASan often, and for a long time. I think this is a good idea, but that it needs to be optional (even if on by default).

Sadly, "false positives are rare[^1]" is not the same as "every program can use ASan without issues". There are a number of situations that are unsupported:

1. Programs with custom C/C++ allocators (e.g. which override malloc/free). ASan needs to override these too, and they can't both do it. Even on macOS where this could actually work because of the zone API, it will still cause issues[^2]
2. Programs that use non-standard system allocator APIs that are not overridden by Asan will often behave badly. Often this is in the form of some custom allocator function asan doesn't happen to have an override for, which is still compatible with `free`. I think this has gotten a lot better, but it seems unlikely to ever be totally fixed.
3. Asan will often reserve a crapload of virutal memory (several TB), which can cause problems on some systems.

[^1]: Honestly, the number of false positives from asan's leak checker is fairly high IME -- at least if you don't build *everything* with ASan on. But false positives for memory safety errors are low, excluding the issues I mention around allocators.

[^2]: For example, jemalloc will [loop until it is not the default zone](https://github.com/aerospike/jemalloc/blob/05108b5010a511226fb7586543f4162dd2d31d2b/src/zone.c#L275). Unfortunately, ASan interposes the `malloc_default_zone` function to always return its zone... which means this loop never terminates.

These are mainly cases that crop up with FFI, but a big reason to use `cargo-careful` is FFI.

Similarly, it may also be worth setting adding `-fsanitize=address` to `CFLAGS` and `CXXFLAGS`, so code built with the e.g. `cc` crate in a build script will get checked too. This may be way more trouble than its worth though, especially since ASan (unlike the other sanitizers) doesn't need it (although it helps the leak check).

File: src/main.rs (hunk 2 above)
Comment (RalfJung) [reply src/main.rs:303]: I have pushed something to master that hopefully makes it clear what I mean.

File: src/main.rs [hunk 4]
```diff
@@ -158,7 +160,67 @@ pub fn ask_to_run(mut cmd: Command, ask: bool, text: &str) {
...
 }
 
-fn build_sysroot(auto: bool, target: &str, rustc_version: &VersionMeta) -> PathBuf {
+/// Returns the sanitizer to use.
+/// Ok(Some(str)) => Use sanitizer "string"
+/// Ok(None) => Use no sanitizer
+///
+/// # Errors
+/// Err(err) => There was an error when trying to get the list of supported sanitizers
+pub fn sanitizer(target: &str) -> anyhow::Result<Option<&'static str>> {
```
Comment (RalfJung) [inline src/main.rs:169]: Please import `anyhow::Result`

File: src/main.rs [hunk 5]
```diff
@@ -158,7 +160,67 @@ pub fn ask_to_run(mut cmd: Command, ask: bool, text: &str) {
...
+    let mut cmd = rustc();
+    cmd.args([
+        "-Z",
+        "unstable-options",
+        "--print",
+        "target-spec-json",
+        "--target",
+        target,
+    ]);
+    let output = match cmd.output() {
```
Comment (RalfJung) [inline src/main.rs:182]: The pattern I usually use is `cmd.output().context("failed to ...")?`. That's much less noisy than explicitly matching. Any reason not to do the same here?

File: src/main.rs [hunk 6]
```diff
@@ -158,7 +160,67 @@ pub fn ask_to_run(mut cmd: Command, ask: bool, text: &str) {
...
+        }
+    };
+
+    // The list of supported sanitizers is stored as an array
+    // in the "supported-sanitizers" key of the target JSON
+    match map.get("supported-sanitizers") {
+        Some(serde_json::Value::Array(arr)) => Ok(arr
+            .contains(&serde_json::Value::String(SANITIZER.to_string()))
+            .then_some(SANITIZER)),
+        Some(_) => Err(anyhow!(
```
Comment (RalfJung) [inline src/main.rs:211]: `bail!` does the same but is shorter

File: src/main.rs [hunk 7]
```diff
@@ -188,14 +250,20 @@ fn build_sysroot(auto: bool, target: &str, rustc_version: &VersionMeta) -> PathB
 
     // Determine where to put the sysroot.
     let user_dirs = directories::ProjectDirs::from("de", "ralfj", "cargo-careful").unwrap();
-    let sysroot_dir = user_dirs.cache_dir();
+    let mut sysroot_dir: PathBuf = user_dirs.cache_dir().to_owned();
 
-    // Do the build.
-    eprint!("Preparing a careful sysroot (target: {target})... ");
+    if let Some(san) = sanitizer {
+        sysroot_dir.push(san);
```
Comment (RalfJung) [inline src/main.rs:256]:
```suggestion
        // Use a separate sysroot dir, to get separate caching of builds with and without sanitizer.
        sysroot_dir.push(san);
```

File: src/main.rs [hunk 8]
```diff
@@ -188,14 +250,20 @@ fn build_sysroot(auto: bool, target: &str, rustc_version: &VersionMeta) -> PathB
...
+        eprint!("Preparing a careful sysroot (target: {target}, sanitizer: {san})... ")
+    } else {
+        eprint!("Preparing a careful sysroot (target: {target})... ")
+    }
     if !auto {
         eprintln!();
     }
-    Sysroot::new(sysroot_dir, target)
+
+    // Do the build.
```
Comment (RalfJung) [inline src/main.rs:265]: I am fine with making this a separate 'paragraph' but then please also add an empty line after the `build_from_source` and before the subsequent printing.

**Review by RalfJung — COMMENTED** [review]
Now I can understand the code, thanks. :) I left some comments.

Please also add a section in the readme explaining the new flag, and what we do with ASAN_OPTIONS.

File: src/main.rs (hunk 4 above)
Comment (Jules-Bertholet) [reply src/main.rs]: I usually prefer leaving this explicit, to differentiate from non-typedef'd `Result`. But I've imported it as you requested

**Comment by Jules-Bertholet** [comment]
I've addressed all the review comments and added a note to the readme.

File: src/main.rs (hunk 4 above)
Comment (RalfJung) [reply src/main.rs]: Yeah this is definitely a matter of personal preference. When there is only one `Result` in use in a project, I prefer to import it.

**Review by RalfJung — COMMENTED** [review]
Looking good, just some nits. :)

Please also squash the commits.

**Comment by Jules-Bertholet** [comment]
It's done.

**Comment by RalfJung** [comment]
CI seems to be [unhappy](https://github.com/RalfJung/cargo-careful/actions/runs/3179162824/jobs/5186044115) on macOS?
```
  = note: ld: library not found for -lrustc-nightly_rt.asan
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by RalfJung** [comment]
@Jules-Bertholet any idea what is going on with the macOS CI failures here? Is that something you plan to look into?

As-is the PR clearly cannot land.

**Comment by Jules-Bertholet** [comment]
@RalfJung It turns out that in addition to the macOS issue (which I haven't looked into), there is another problem: the `-Zsanitizer` feature doesn't play well with proc macros. Until that rustc issue is fixed, sanitizing can't really be enabled by default. I could modify this PR to make sanitizing off-by-default, if you prefer.

**Comment by RalfJung** [comment]
Ah, yeah we don't have any proc macros in the test suite here I think.

An off-by-default flag sounds reasonable, with the docs saying it is currently experimental and not tested on CI and pointing to the rust issue about proc macros.

**Comment by Jules-Bertholet** [comment]
Ready for review

File: README.md [hunk 9]
```diff
@@ -47,3 +49,21 @@ That said, there is a lot of Undefined Behavior that is *not* detected by `cargo
...
+To use a sanitizer, pass the command-line flag `-Zcareful-sanitizer=<your_sanitizer>` to `cargo careful`.
+The list of supported sanitizers and targets can be found
+[here](https://doc.rust-lang.org/nightly/unstable-book/compiler-flags/sanitizer.html).
+If you pass `-Zcareful-sanitizer` without specifying a sanitizer, [`AddressSanitizer`](https://clang.llvm.org/docs/AddressSanitizer.html)
+will be used.
+
+By default, when using `AddressSanitizer`, `cargo careful` will disable memory leak checking by
+setting `ASAN_OPTIONS=detect_leaks=0` in your program's environment, as memory leaks are not
+usually a soundness or correctness issue. If you set the `ASAN_OPTIONS` environment variable
+yourself, that will override this behavior.
```
Comment (RalfJung) [inline README.md:69]:
```suggestion
yourself (e.g. to an empty value), that will override this behavior.
```

File: src/main.rs [hunk 10]
```diff
@@ -15,6 +16,7 @@ const CAREFUL_FLAGS: &[&str] = &[
     "-Cdebug-assertions=on",
 ];
 const STD_FEATURES: &[&str] = &["panic_unwind", "backtrace"];
+const DEFAULT_SANITIZER: &str = "address";
```
Comment (RalfJung) [inline src/main.rs:19]:
```suggestion
/// The sanitizer to use when just `-Zcareful-sanitizer` is passed as flag.
const DEFAULT_SANITIZER: &str = "address";
```

File: src/main.rs [hunk 11]
```diff
@@ -239,25 +305,61 @@ fn cargo_careful(mut args: env::Args) {
...
+                None
+            },
+            |b| {
+                if b {
+                    eprintln!(
+                        "Using sanitizier `{san}`."
+                    );
+                    Some(san)
+                } else {
+                    eprintln!("Sanitizer `{san}` not supported by target, falling back to none.");
```
Comment (RalfJung) [inline src/main.rs:349]: Wouldn't it be better to bail out here? Seems like an error if the user asks for a sanitizer that is not supported.

File: src/main.rs [hunk 12]
```diff
@@ -158,7 +160,58 @@ pub fn ask_to_run(mut cmd: Command, ask: bool, text: &str) {
...
 }
 
-fn build_sysroot(auto: bool, target: &str, rustc_version: &VersionMeta) -> PathBuf {
+/// Returns the sanitizer to use.
+/// Ok(true) => Use sanitizer "string"
+/// Ok(false) => Use no sanitizer
+///
+/// # Errors
+/// Err(err) => There was an error when trying to get the list of supported sanitizers
+pub fn sanitizer(target: &str, san: &str) -> Result<bool> {
```
Comment (RalfJung) [inline src/main.rs:169]:
```suggestion
/// Returns whether the given sanitizer is supported.
///
/// # Errors
/// Err(err) => There was an error when trying to get the list of supported sanitizers
pub fn is_sanitizer_supported(target: &str, san: &str) -> Result<bool> {
```
Is this correct? The old doc comment confused me.

**Comment by RalfJung** [comment]
Looks good! Just some minor nits, and there are conflicts.

**Comment by Jules-Bertholet** [comment]
Should be good

File: src/main.rs [hunk 13]
```diff
@@ -159,7 +163,56 @@ pub fn ask_to_run(mut cmd: Command, ask: bool, text: &str) {
     }
 }
 
-fn build_sysroot(auto: bool, target: &str, rustc_version: &VersionMeta) -> PathBuf {
+/// Returns whether the given sanitizer is supported on this target.
+///
+/// # Errors
+/// Returns `Err` if there was an error when geutting the list of supported sanitizers.
```
Comment (RalfJung) [inline src/main.rs:169]:
```suggestion
/// Returns `Err` if there was an error when getting the list of supported sanitizers.
```

**Pull request merged**